File: python/code/7-find-spectrum.py
import numpy as np
import matplotlib.pyplot as plt

# Number of samples
N = 256

# Sample index
k = np.arange(N)

# Given signal
x = (
    0.25
    + 2 * np.sin(2 * np.pi * 5 * k / N)
    + np.sin(2 * np.pi * 12.5 * k / N)
    + 1.5 * np.sin(2 * np.pi * 20 * k / N)
    + 0.5 * np.sin(2 * np.pi * 35 * k / N)
)

# FFT
X = np.fft.fft(x)

# Frequency
freq = np.fft.fftfreq(N)

# Magnitude
magnitude = np.abs(X) / N

# Plot positive frequencies
positive = freq >= 0

plt.figure(figsize=(10, 5))
plt.stem(freq[positive], magnitude[positive])

plt.title("Spectrum of Given Signal")
plt.xlabel("Normalized Frequency")
plt.ylabel("Magnitude")
plt.grid(True)

plt.show()

File: matlab/code/7-find-spectrum.m
clc;
clear;
close all;

% Number of samples
N = 1000;

% Time/sample variable
k = (0:N-1) / N;

% Given signal
x = 0.25 ...
    + 2*sin(2*pi*5*k) ...
    + sin(2*pi*12.5*k) ...
    + 1.5*sin(2*pi*20*k) ...
    + 0.5*sin(2*pi*35*k);

% FFT
X = fft(x);

% Magnitude
X_mag = abs(X) / N;

% Frequency axis
f = (0:N-1);

% Plot signal
figure;
plot(k, x);
grid on;
xlabel('k');
ylabel('Amplitude');
title('Given Signal');

% Plot spectrum
figure;
plot(f(1:N/2), 2*X_mag(1:N/2));
grid on;
xlabel('Frequency');
ylabel('Magnitude');
title('Spectrum of the Given Signal');
xlim([0 50]);